File: MATEMÁTICA/1393/main.c
#include <stdio.h>

int fibonacci(int n, int memo[42]) {
    if (memo[n]) return memo[n];
    memo[n] = n < 2 ? n : fibonacci(n - 1, memo) + fibonacci(n - 2, memo);
    return memo[n];
}

int main() {
    int n, memo[42] = {0};
    scanf("%d", &n);
    while (n) {
        printf("%d\n", fibonacci(n + 1, memo));
        scanf("%d", &n);
    }
    return 0;
}

File: MATEMÁTICA/1393/main.rb
def fibonacci(n, memo)
  return memo[n] if memo[n].nonzero?
  memo[n] = (n < 2) ? n : fibonacci(n - 1, memo) + fibonacci(n - 2, memo)
  memo[n]
end

memo = [0] * 42
while (n = gets.to_i).nonzero?
  puts fibonacci(n + 1, memo)
end
